consensus/src/chained_bft/block_storage/pending_votes.rs

```
@@ -0,0 +1,166 @@
+        tc.add_signature(author, vote_msg.signature().clone());
+        if tc.signatures().len() >= min_votes_for_tc {
+            return Some(VoteReceptionResult::NewTimeoutCertificate(Arc::new(
+                tc.clone(),
+            )));
+        }
+        None
+    }
+
+    /// Check if vote is valid. If this is the first vote from Author, add it to map. If Author has
+    /// already voted on same block then return DuplicateVote error. If Author has already voted
+    /// on some other result, prune the last vote and insert new one in map.
+    fn check_vote_valid(&mut self, vote_msg: &VoteMsg) -> Result<(), VoteReceptionResult> {
+        let author = vote_msg.author();
+        let round = vote_msg.vote_data().block_round();
+        let li_digest = vote_msg.ledger_info().hash();
+        let vote_info = LastVoteInfo { li_digest, round };
+        let last_voted_info = match self.author_to_last_voted_info.insert(author, vote_info) {
+            None => {
+                // First vote from Author, do nothing.
+                return Ok(());
+            }
+            Some(last_voted_info) => last_voted_info,
+        };
+
+        if li_digest == last_voted_info.li_digest {
+            // Author has already voted for this LedgerInfo or round
+            return Err(VoteReceptionResult::DuplicateVote);
+        }
+        // Prune last pending vote from the pending votes.
+        if let Some(li_pending_votes) = self.li_digest_to_votes.get_mut(&last_voted_info.li_digest)
+        {
+            // Removing signature from last voted block
+            li_pending_votes.remove_signature(author);
+            if li_pending_votes.signatures().is_empty() {
+                // Last vote for that LI digest, remove the digest entry
+                self.li_digest_to_votes.remove(&last_voted_info.li_digest);
+            }
+        }
+
+        // Prune last pending vote from the pending timeout certificates.
+        if round != last_voted_info.round {
```

why we have this check? duplicate vote should be checked in line 138 already?
if we have a vote for same round but different li_digest we probably want to log it and return error instead of ignoring it

---

I'm going to extend the VoteReceptionResult enum with a new value for EquivocateVote to be returned to the user (we couldn't detect it previously but now we can).